### PR TITLE
Quern

### PR DESCRIPTION
**Pull request by CtrlAltDavid01**
- Rebased quern
- Modified the recipes to fit with the system in place
- Quite a few performance improvements

### CODE REVIEW

File: src/main/java/net/dries007/tfc/objects/blocks/devices/BlockQuern.java [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+                    if (hitX > 0.2f && hitX < 0.8f && hitZ > 0.2f && hitZ < 0.8f)
+                    {
+                        if (hitX > 0.4f && hitX < 0.6f && hitZ > 0.4f && hitZ < 0.6f && playerStack.isEmpty() && !inputStack.isEmpty())
+                        {
+                            playerIn.setHeldItem(EnumHand.MAIN_HAND, inventory.extractItem(TEQuern.SLOT_INPUT, inputStack.getCount(), false));
+                            teQuern.setAndUpdateSlots(TEQuern.SLOT_INPUT);
+                            world.playSound(null, pos, SoundEvents.BLOCK_CLOTH_HIT, SoundCategory.BLOCKS, 1, 1);
+                        }
+                        //Can't put an item in there because of "doesSneakBypassUse" is false on the items that can be ground
+                        /*else if (hitX > 0.4f && hitX < 0.6f && hitZ > 0.4f && hitZ < 0.6f && !playerStack.isEmpty() && isValid)
```
Comment (alcatrazEscapee) [inline src/main/java/net/dries007/tfc/objects/blocks/devices/BlockQuern.java]: If this is not needed anymore, remove it. Code comments for things that don't work generally don't add anything documentation wise.

File: src/main/java/net/dries007/tfc/objects/inventory/slot/SlotQuernHandstone.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
+/*
```
Comment (alcatrazEscapee) [inline src/main/java/net/dries007/tfc/objects/inventory/slot/SlotQuernHandstone.java]: None of this is necessary, all the functionality should be able to be handled by `SlotCallback`

File: src/main/java/net/dries007/tfc/objects/inventory/slot/SlotQuernInput.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
+package net.dries007.tfc.objects.inventory.slot;
```
Comment (alcatrazEscapee) [inline src/main/java/net/dries007/tfc/objects/inventory/slot/SlotQuernInput.java]: Same as above, all of this should be done with `SlotCallback`

File: src/main/java/net/dries007/tfc/objects/inventory/slot/SlotQuernOutput.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
+/*
```
Comment (alcatrazEscapee) [inline src/main/java/net/dries007/tfc/objects/inventory/slot/SlotQuernOutput.java]: Same as above, all of this should be done with `SlotCallback`. As for the xp, I'm thinking we should just have the quern not produce xp for now. Reason being: no other devices produce xp at the moment, and I'd rather keep things consistent for the time being.

File: src/main/java/net/dries007/tfc/objects/recipes/QuernRecipe.java [hunk 5]
```diff
@@ -0,0 +1,109 @@
+/*
```
Comment (alcatrazEscapee) [inline src/main/java/net/dries007/tfc/objects/recipes/QuernRecipe.java]: Remove this.

File: src/main/java/net/dries007/tfc/objects/recipes/QuernRecipeManager.java [hunk 6]
```diff
@@ -0,0 +1,134 @@
+/*
```
Comment (alcatrazEscapee) [inline src/main/java/net/dries007/tfc/objects/recipes/QuernRecipeManager.java]: Remove this

File: src/main/java/net/dries007/tfc/objects/te/TEQuern.java [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+
+        if (quernStack.isEmpty() || (playerStack.isStackable() && quernStack.isStackable() && quernStack.getItem() == playerStack.getItem() && (!playerStack.getHasSubtypes() || playerStack.getMetadata() == quernStack.getMetadata()) && ItemStack.areItemStackTagsEqual(playerStack, quernStack)))
+        {
+            return inventory.insertItem(slot, playerStack, false);
+        }
+        inventory.setStackInSlot(slot, playerStack);
+        return quernStack;
+    }
+
+    public ItemStackHandler getInventory()
```
Comment (alcatrazEscapee) [inline src/main/java/net/dries007/tfc/objects/te/TEQuern.java]: Anything that calls this should instead call `tile.getCapability(CapabilityItemHandler.ITEM_HANDLER_CAPABILITY, null)`, and then null check the capability. This method should be removed

File: src/main/java/net/dries007/tfc/objects/te/TEQuern.java [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+    {
+        return inventory;
+    }
+
+    public void setInventory(@Nonnull NBTTagCompound nbt)
```
Comment (alcatrazEscapee) [inline src/main/java/net/dries007/tfc/objects/te/TEQuern.java]: Lots of this stuff can be removed - it was old and used by old systems

File: src/main/java/net/dries007/tfc/client/sound/TFCSounds.java [hunk 9]
```diff
@@ -0,0 +1,39 @@
+/*
```
[the remark this answers is not in the document]
Comment (CtrlAltDavid01) [reply src/main/java/net/dries007/tfc/client/sound/TFCSounds.java]: I deleted it, I don't know how it came back, I'll try to delete it again

File: src/main/java/net/dries007/tfc/network/PacketQuernUpdate.java [hunk 10]
```diff
@@ -0,0 +1,70 @@
+/*
```
[the remark this answers is not in the document]
Comment (CtrlAltDavid01) [reply src/main/java/net/dries007/tfc/network/PacketQuernUpdate.java]: I deleted this too, will try again

File: src/main/java/net/dries007/tfc/network/PacketRequestQuernUpdate.java [hunk 11]
```diff
@@ -0,0 +1,71 @@
+/*
```
[the remark this answers is not in the document]
Comment (CtrlAltDavid01) [reply src/main/java/net/dries007/tfc/network/PacketRequestQuernUpdate.java]: Deleted this as well...

File: src/main/java/net/dries007/tfc/objects/recipes/QuernRecipeManager.java (hunk 6 above)
Comment (CtrlAltDavid01) [reply src/main/java/net/dries007/tfc/objects/recipes/QuernRecipeManager.java]: I don't know why the deletes didn't work

File: src/main/java/net/dries007/tfc/types/DefaultRecipes.java [hunk 12]
```diff
@@ -275,6 +277,57 @@ public static void onRegisterLoomRecipeEvent(RegistryEvent.Register<LoomRecipe>
...
 
+    @SubscribeEvent
+    public static void onRegisterQuernRecipeEvent(RegistryEvent.Register<QuernRecipe> event)
+    {
+        IForgeRegistry<QuernRecipe> r = event.getRegistry();
+
+        r.registerAll(
+            //Flux
+            new QuernRecipe(new ResourceLocation(MOD_ID, "boarx"), IIngredient.of("gemBorax"), new ItemStack(ItemPowder.get(Powder.FLUX), 6)),
+            new QuernRecipe(new ResourceLocation(MOD_ID, "chalk"), IIngredient.of("rockChalk"), new ItemStack(ItemPowder.get(Powder.FLUX), 2)),
```
Comment (alcatrazEscapee) [inline src/main/java/net/dries007/tfc/types/DefaultRecipes.java]: Minor nitpick: these four recipes can be one, and use the existing ore dict name `rockFlux`

File: src/main/java/net/dries007/tfc/objects/te/TEQuern.java [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+                    world.playSound(null, pos, ENTITY_ITEM_BREAK, SoundCategory.BLOCKS, 1.0f, 0.6f);
+                }
+
+                setAndUpdateSlots(SLOT_HANDSTONE);
+            }
+        }
+    }
+
+    @Override
+    public boolean shouldRefresh(World world, BlockPos pos, IBlockState oldState, IBlockState newSate)
```
Comment (alcatrazEscapee) [inline src/main/java/net/dries007/tfc/objects/te/TEQuern.java]: Not nessecary, already included in `TEInventory` via parent class `TEBase`

**Pull request merged**